File: common/src/main/java/io/xol/chunkstories/world/region/HashMapWorldRegionsHolder.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.world.region;

import java.util.*;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

import io.xol.chunkstories.api.util.concurrency.Fence;
import io.xol.chunkstories.api.world.WorldUser;
import io.xol.chunkstories.api.world.chunk.Chunk;
import io.xol.chunkstories.api.world.chunk.ChunkHolder;
import io.xol.chunkstories.api.world.region.Region;
import io.xol.chunkstories.util.concurrency.CompoundFence;
import io.xol.chunkstories.world.WorldImplementation;
import io.xol.chunkstories.world.chunk.CubicChunk;

public class HashMapWorldRegionsHolder {
	private WorldImplementation world;

	// private Semaphore noConcurrentRegionCreationDestruction = new Semaphore(1);
	private final ReadWriteLock regionsLock = new ReentrantReadWriteLock();
	private final Map<Integer, RegionImplementation> regions = new HashMap<>();

	private final int sizeInRegions, heightInRegions;

	public HashMapWorldRegionsHolder(WorldImplementation world) {
		this.world = world;
		// this.chunksData = chunksData;
		heightInRegions = world.getWorldInfo().getSize().heightInChunks / 8;
		sizeInRegions = world.getWorldInfo().getSize().sizeInChunks / 8;
	}

	//TODO have a copyOnWrite list or smth
	public Collection<RegionImplementation> internalGetLoadedRegions() {
		try {
			regionsLock.readLock().lock();
			List<RegionImplementation> list = new ArrayList<>();
			list.addAll(regions.values());
			return list;
		} finally {
			regionsLock.readLock().unlock();
		}
	}

	public RegionImplementation getRegionChunkCoordinates(int chunkX, int chunkY, int chunkZ) {
		return getRegion(chunkX / 8, chunkY / 8, chunkZ / 8);
	}

	public RegionImplementation getRegion(int regionX, int regionY, int regionZ) {
		try {
			regionsLock.readLock().lock();
			int key = (regionX * sizeInRegions + regionZ) * heightInRegions + regionY;
			return regions.get(key);
		} finally {
			regionsLock.readLock().unlock();
		}
	}

	long prout = 0;

	public CubicChunk getChunk(int chunkX, int chunkY, int chunkZ) {
		RegionImplementation holder = getRegionChunkCoordinates(chunkX, chunkY, chunkZ);
		if (holder != null) {
			// System.out.println(regions.size());
			return holder.getChunk(chunkX, chunkY, chunkZ);
		}
		return null;
	}

	public Fence saveAll() {
		CompoundFence allRegionsFences = new CompoundFence();

		Iterator<RegionImplementation> i = regions.values().iterator();
		Region region;
		while (i.hasNext()) {
			region = i.next();
			if (region != null) {
				allRegionsFences.add(region.save());
			}
		}

		return allRegionsFences;
	}

	/*public void markChunkForReRender(int chunkX, int chunkY, int chunkZ) {
		int worldSizeInChunks = world.getWorldInfo().getSize().sizeInChunks;
		if (chunkX < 0)
			chunkX += worldSizeInChunks;
		if (chunkY < 0)
			chunkY += worldSizeInChunks;
		if (chunkZ < 0)
			chunkZ += worldSizeInChunks;
		chunkX = chunkX % worldSizeInChunks;
		chunkZ = chunkZ % worldSizeInChunks;

		if (chunkY < 0 || chunkY >= worldSizeInChunks)
			return;

		Chunk c = getChunk(chunkX, chunkY, chunkZ);
		if (c != null)
			c.mesh().incrementPendingUpdates();
	}*/

	public void destroy() {
		regions.clear();
	}

	@Override
	public String toString() {
		return "[ChunksHolder: " + regions.size() + " Chunk Holders loaded]";
	}

	public String getStats() {
		return countChunks() + " (lr: " + regions.size() + " )";
	}

	public int countChunks() {
		int c = 0;
		Iterator<Chunk> i = world.getAllLoadedChunks();
		while (i.hasNext()) {
			i.next();
			c++;
		}
		return c;
	}

	public int countChunksWithData() {
		int c = 0;
		Iterator<Chunk> i = world.getAllLoadedChunks();
		while (i.hasNext()) {
			if (!i.next().isAirChunk())
				c++;
		}
		return c;
	}

	/**
	 * Atomically adds an user to a region itself, and creates it if it was
	 * previously unused
	 */
	public RegionImplementation acquireRegion(WorldUser user, int regionX, int regionY, int regionZ) {
		if (regionY < 0 || regionY > world.getMaxHeight() / 256)
			return null;

		int key = (regionX * sizeInRegions + regionZ) * heightInRegions + regionY;

		this.regionsLock.writeLock().lock();

		RegionImplementation region = regions.get(key);
		boolean fresh = false;
		if (region == null) {
			region = new RegionImplementation(world, regionX, regionY, regionZ);
			fresh = true;
		}

		boolean userAdded = region.registerUser(user);

		if (fresh)
			regions.put(key, region);

		this.regionsLock.writeLock().unlock();

		return userAdded ? region : null;
	}

	/**
	 * Atomically adds an user to a region's chunk, and creates the region if it was
	 * previously unused
	 */
	public ChunkHolder acquireChunkHolder(WorldUser user, int chunkX, int chunkY, int chunkZ) {
		if (chunkY < 0 || chunkY > world.getMaxHeight() / 32)
			return null;

		int regionX = chunkX >> 3;
		int regionY = chunkY >> 3;
		int regionZ = chunkZ >> 3;
		int key = (regionX * sizeInRegions + regionZ) * heightInRegions + regionY;

		this.regionsLock.writeLock().lock();

		RegionImplementation region = regions.get(key);
		boolean fresh = false;
		if (region == null) {
			region = new RegionImplementation(world, regionX, regionY, regionZ);
			fresh = true;
		}

		ChunkHolder chunkHolder = region.getChunkHolder(chunkX, chunkY, chunkZ);
		boolean userAdded = chunkHolder.registerUser(user);

		if (fresh)
			regions.put(key, region);

		this.regionsLock.writeLock().unlock();

		return userAdded ? chunkHolder : chunkHolder;
	}

	/**
	 * Callback by the holder's unload() method to remove himself from this list.
	 */
	void removeRegion(RegionImplementation region) {
		this.regionsLock.writeLock().lock();
		int key = (region.getRegionX() * sizeInRegions + region.getRegionZ()) * heightInRegions + region.getRegionY();
		regions.remove(key);
		this.regionsLock.writeLock().unlock();
	}
}

File: common/src/main/java/io/xol/chunkstories/world/region/HashMapWorldRegionsHolder.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.world.region

import java.util.*
import java.util.concurrent.locks.ReadWriteLock
import java.util.concurrent.locks.ReentrantReadWriteLock

import io.xol.chunkstories.api.util.concurrency.Fence
import io.xol.chunkstories.api.world.WorldUser
import io.xol.chunkstories.api.world.chunk.ChunkHolder
import io.xol.chunkstories.util.concurrency.CompoundFence
import io.xol.chunkstories.world.WorldImplementation
import io.xol.chunkstories.world.chunk.CubicChunk

class HashMapWorldRegionsHolder(private val world: WorldImplementation) {

    private val regionsLock = ReentrantReadWriteLock()

    private val regions: MutableMap<Int, RegionImplementation> = mutableMapOf()
    private var regionsList: List<RegionImplementation> = emptyList()

    private val sizeInRegions = world.worldInfo.size.sizeInChunks / 8
    private val heightInRegions = world.worldInfo.size.heightInChunks / 8

    val stats: String
        get() = countChunks().toString() + " (lr: " + regions.size + " )"

    //TODO have a copyOnWrite list or smth
    fun internalGetLoadedRegions(): Collection<RegionImplementation> = regionsList
    /*{
        try {
            regionsLock.readLock().lock()
            val list = ArrayList<RegionImplementation>()
            list.addAll(regions.values)
            return list
        } finally {
            regionsLock.readLock().unlock()
        }
    }*/

    fun internalGetLoadedChunks(): Collection<CubicChunk> = internalGetLoadedRegions().flatMap { it.loadedChunks }

    fun getRegionChunkCoordinates(chunkX: Int, chunkY: Int, chunkZ: Int): RegionImplementation? {
        return getRegion(chunkX / 8, chunkY / 8, chunkZ / 8)
    }

    fun getRegion(regionX: Int, regionY: Int, regionZ: Int): RegionImplementation? {
        try {
            regionsLock.readLock().lock()
            val key = (regionX * sizeInRegions + regionZ) * heightInRegions + regionY
            return regions[key]
        } finally {
            regionsLock.readLock().unlock()
        }
    }

    fun getChunk(chunkX: Int, chunkY: Int, chunkZ: Int): CubicChunk? {
        val holder = getRegionChunkCoordinates(chunkX, chunkY, chunkZ)
        return holder?.getChunk(chunkX, chunkY, chunkZ)
    }

    fun saveAll(): Fence {
        val allRegionsFences = CompoundFence()
        try {
            regionsLock.readLock().lock()
            regionsList.mapTo(allRegionsFences) { it.save() }
            return allRegionsFences
        } finally {
            regionsLock.readLock().unlock()
        }
    }

    fun destroy() {
        regions.clear()
    }

    override fun toString(): String {
        return "[RegionsHolder: " + regions.size + " loaded regions]"
    }

    fun countChunks() : Int = regionsList.size

    /**
     * Atomically adds an user to a region itself, and creates it if it was
     * previously unused
     */
    fun acquireRegion(user: WorldUser, regionX: Int, regionY: Int, regionZ: Int): RegionImplementation? {
        if (regionY < 0 || regionY > world.maxHeight / 256)
            return null

        val key = (regionX * sizeInRegions + regionZ) * heightInRegions + regionY

        this.regionsLock.writeLock().lock()

        var region: RegionImplementation? = regions[key]
        var fresh = false
        if (region == null) {
            region = RegionImplementation(world, regionX, regionY, regionZ)
            fresh = true
        }

        val userAdded = region.registerUser(user)

        if (fresh) {
            regions[key] = region
            regionsList = regions.values.toList()
        }

        this.regionsLock.writeLock().unlock()

        return if (userAdded) region else null
    }

    /**
     * Atomically adds an user to a region's chunk, and creates the region if it was
     * previously unused
     */
    fun acquireChunkHolder(user: WorldUser, chunkX: Int, chunkY: Int, chunkZ: Int): ChunkHolder? {
        if (chunkY < 0 || chunkY > world.maxHeight / 32)
            return null

        val regionX = chunkX shr 3
        val regionY = chunkY shr 3
        val regionZ = chunkZ shr 3
        val key = (regionX * sizeInRegions + regionZ) * heightInRegions + regionY

        this.regionsLock.writeLock().lock()

        var region: RegionImplementation? = regions[key]
        var fresh = false
        if (region == null) {
            region = RegionImplementation(world, regionX, regionY, regionZ)
            fresh = true
        }

        val chunkHolder = region.getChunkHolder(chunkX, chunkY, chunkZ)
        val userAdded = chunkHolder.registerUser(user)

        if (fresh) {
            regions[key] = region
            regionsList = regions.values.toList()
        }

        this.regionsLock.writeLock().unlock()

        return if (userAdded) chunkHolder else chunkHolder
    }

    /**
     * Callback by the holder's unload() method to remove himself from this list.
     */
    fun removeRegion(region: RegionImplementation) {
        this.regionsLock.writeLock().lock()
        val key = (region.getRegionX() * sizeInRegions + region.getRegionZ()) * heightInRegions + region.getRegionY()
        regions.remove(key)
        regionsList = regions.values.toList()
        this.regionsLock.writeLock().unlock()
    }
}
